python/pyspark/shuffle.py

```diff
@@ -529,6 +520,295 @@ def sorted(self, iterator, key=None, reverse=False):
         return heapq.merge(chunks, key=key, reverse=reverse)
 
 
+class ExternalList(object):
+    """
+    ExternalList can have many items which cannot be hold in memory in
+    the same time.
+
+    >>> l = ExternalList(range(100))
+    >>> len(l)
+    100
+    >>> l.append(10)
+    >>> len(l)
+    101
+    >>> for i in range(10240):
+    ...     l.append(i)
+    >>> len(l)
+    10341
+    >>> import pickle
+    >>> l2 = pickle.loads(pickle.dumps(l))
+    >>> len(l2)
+    10341
+    >>> list(l2)[100]
+    10
+    """
+    LIMIT = 10240
+
+    def __init__(self, values):
+        self.values = values
+        self.disk_count = 0
+        self._file = None
+        self._ser = None
+
+    def __getstate__(self):
+        if self._file is not None:
+            self._file.flush()
+            f = os.fdopen(os.dup(self._file.fileno()))
+            f.seek(0)
+            bytes = f.read()
+        else:
+            bytes = ''
+        return self.values, self.disk_count, bytes
+
+    def __setstate__(self, item):
+        self.values, self.disk_count, bytes = item
+        if bytes:
+            self._open_file()
+            self._file.write(bytes)
+        else:
+            self._file = None
+            self._ser = None
+
+    def __iter__(self):
+        if self._file is not None:
+            self._file.flush()
+            # read all items from disks first
+            with os.fdopen(os.dup(self._file.fileno()), 'r') as f:
+                f.seek(0)
+                for values in self._ser.load_stream(f):
+                    for v in values:
+                        yield v
+
+        for v in self.values:
+            yield v
+
+    def __len__(self):
+        return self.disk_count + len(self.values)
+
+    def append(self, value):
+        self.values.append(value)
+        # dump them into disk if the key is huge
+        if len(self.values) >= self.LIMIT:
+            self._spill()
+
+    def _open_file(self):
+        dirs = _get_local_dirs("objects")
+        d = dirs[id(self) % len(dirs)]
+        if not os.path.exists(d):
+            os.makedirs(d)
+        p = os.path.join(d, str(id))
+        self._file = open(p, "w+", 65536)
+        self._ser = CompressedSerializer(PickleSerializer())
+        os.unlink(p)
```

Comment:
Why do we need this `unlink` call here?  Is this to delete the file if it already exists?